app/controllers/carts_controller.rb

```diff
@@ -1,24 +1,41 @@
 class CartsController < ApplicationController
+  before_action :find_current_cart
+  before_action :find_cart_products, only: :index
+  before_action :update_price, only: :index
+
   def index
     redirect_to root_path unless logged_in?
-
-    @cart = session[:cart]
-    @price = update_price
   end
 
   def update_price
     price = 0
-    return price if @cart.nil?
+    return price if @current_cart.nil?
 
-    @cart.each do |index, item|
-      product = Product.find(index)
-      price += product.price * item['quantity']
+    @current_cart.cart_items.each do |prod|
```

Comment:
I think you should rename this `prod` variable. First of all, I'd avoid abbreviations for words that aren't very long; second, it gives me the impression that the class is a `Product` and that the `Product` class has a field `product_id`, which would be redundant and confusing.

You could try something like `cart_item` or `item`.